versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java

```diff
@@ -0,0 +1,265 @@
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    databaseAdapter.assign(ref, expectedHash, targetHash);
+  }
+
+  @Override
+  public Hash create(NamedRef ref, Optional<Hash> targetHash)
+      throws ReferenceNotFoundException, ReferenceAlreadyExistsException {
+    if (ref instanceof TagName && !targetHash.isPresent()) {
+      throw new IllegalArgumentException(
+          "Tag-creation requires a target named-reference and hash.");
+    }
+
+    return databaseAdapter.create(ref, targetHash.orElseGet(databaseAdapter::noAncestorHash));
+  }
+
+  @Override
+  public void delete(NamedRef ref, Optional<Hash> hash)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    databaseAdapter.delete(ref, hash);
+  }
+
+  @Override
+  public Stream<WithHash<NamedRef>> getNamedRefs() {
+    return databaseAdapter.namedRefs();
+  }
+
+  @Override
+  public Stream<WithHash<METADATA>> getCommits(Ref ref) throws ReferenceNotFoundException {
+    Hash hash = ref instanceof NamedRef ? toHash((NamedRef) ref) : (Hash) ref;
+    Stream<CommitLogEntry> stream = databaseAdapter.commitLog(hash);
+
+    return stream.map(
+        e ->
+            WithHash.of(
+                e.getHash(), storeWorker.getMetadataSerializer().fromBytes(e.getMetadata())));
+  }
+
+  @Override
+  public Stream<WithType<Key, CONTENTS_TYPE>> getKeys(Ref ref) throws ReferenceNotFoundException {
+    Hash hash = ref instanceof NamedRef ? toHash((NamedRef) ref) : (Hash) ref;
+    return databaseAdapter
+        .keys(hash, KeyFilterPredicate.ALLOW_ALL)
+        .map(kt -> WithType.of(storeWorker.getType(kt.getType()), kt.getKey()));
+  }
+
+  @Override
+  public CONTENTS getValue(Ref ref, Key key) throws ReferenceNotFoundException {
+    return getValues(ref, Collections.singletonList(key)).get(0).orElse(null);
+  }
+
+  @Override
+  public List<Optional<CONTENTS>> getValues(Ref ref, List<Key> keys)
+      throws ReferenceNotFoundException {
+    Hash hash = ref instanceof NamedRef ? toHash((NamedRef) ref) : (Hash) ref;
+    try (Stream<Optional<CONTENTS>> values =
+        databaseAdapter
+            .values(hash, keys, KeyFilterPredicate.ALLOW_ALL)
+            .map(contentsAndStateOptional -> contentsAndStateOptional.map(mapContentsAndState()))) {
+      return values.collect(Collectors.toList());
+    }
+  }
+
+  private Function<ContentsAndState<ByteString>, CONTENTS> mapContentsAndState() {
+    return cs ->
+        storeWorker.valueFromStore(cs.getRefState(), Optional.ofNullable(cs.getGlobalState()));
+  }
+
+  @Override
+  public Stream<Diff<CONTENTS>> getDiffs(Ref from, Ref to) throws ReferenceNotFoundException {
+    Hash fromHash = from instanceof NamedRef ? toHash((NamedRef) from) : (Hash) from;
+    Hash toHash = to instanceof NamedRef ? toHash((NamedRef) to) : (Hash) to;
+    return databaseAdapter
+        .diff(fromHash, toHash, KeyFilterPredicate.ALLOW_ALL)
```

Comment:
If number of diffs are large, can this call cause a heap pressure?

Reply:
Many keys/values can do. See `org.projectnessie.versioned.persist.adapter.spi.AbstractDatabaseAdapter#buildDiff`
But there's room for improvement to not eagerly collect everything but use `Stream`s more intensively. Not as easy as it sounds, because that should not lead for example in fetching each global state individually (see AbstractDatabaseAdapter#fetchValues).

I don't think that it will become an issue soon. Let's not prematurely optimize things.